packages/build-tools/src/utils/packageManager.ts

```diff
@@ -34,3 +35,59 @@ export async function isAtLeastNpm7Async(): Promise<boolean> {
   const version = (await spawnAsync('npm', ['--version'], { stdio: 'pipe' })).stdout.trim();
   return semver.gte(version, '7.0.0');
 }
+
+export function shouldUseFrozenLockfile({
+  env,
+  sdkVersion,
+  reactNativeVersion,
+}: {
+  env: Record<string, string | undefined>;
+  sdkVersion: string | undefined;
+  reactNativeVersion: string | undefined;
+}): boolean {
+  if (env.EAS_NO_FROZEN_LOCKFILE) {
+    return false;
+  }
+
+  if (sdkVersion && semver.lt(sdkVersion, '53.0.0')) {
+    // Before SDK 53 we could not have used frozen lockfile.
+    return false;
+  }
+
+  if (reactNativeVersion && semver.lt(reactNativeVersion, '0.79.0')) {
+    // Before react-native 0.79 we could not have used frozen lockfile.
+    return false;
+  }
```

Comment:
not a blocker here but the dependence on reading a version from package.json feels like a bit of a ticking timebomb for us - at some point someone will need to use a fork and anything that depends on this type of logic will have undefined behavior. we may want to proactively figure out an alternative way to support this -- so if we detect you are using a fork of either package in your package.json then we let people know in eas cli or elsewhere that they need to specify the version explicitly for either or both of the packages somewhere in their package.json

Reply:
Totally agree. Added https://linear.app/expo/issue/ENG-15625/review-package-version-dependant-code-in-worker-and-think-of-better